lib/Dialect/FIRRTL/Transforms/InferReadWrite.cpp

```diff
@@ -0,0 +1,268 @@
+            if (auto sf = dyn_cast<SubfieldOp>(u)) {
+              auto fName =
+                  sf.input().getType().cast<BundleType>().getElementName(
+                      sf.fieldIndex());
+              if (fName.equals("en")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(rEnWire);
+                else
+                  sf->replaceAllUsesWith(wEnWire);
+              } else if (fName.equals("clk")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(clk);
+                else {
+                  auto dummyWire = builder.create<WireOp>(sf.getType());
+                  sf->replaceAllUsesWith(dummyWire);
+                }
+              } else if (fName.equals("addr")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(rAddr);
+                else
+                  sf->replaceAllUsesWith(wAddr);
+              } else if (fName.equals("data")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(readData);
+                else
+                  sf->replaceAllUsesWith(writeData);
+              } else if (fName.equals("mask")) {
+                if (!readPort)
+                  sf->replaceAllUsesWith(mask);
+              }
+              // Once all the uses of the subfield op replaced, delete it.
+              sf.erase();
+            }
+        }
+        // All uses for all results of mem removed, now erase the mem.
+        mem.erase();
+      }
+    }
+  }
+
+private:
+  Value getConnectSrc(Value dst) {
+    for (auto c : dst.getUsers())
+      if (auto connect = dyn_cast<ConnectOp>(c))
+        if (connect.dest() == dst)
+          return connect.src();
+
+    return nullptr;
+  }
+
+  void getProductTerms(Value a, SmallVector<Value> &terms) {
+    if (!a)
+      return;
+    terms.push_back(a);
+    if (a.isa<BlockArgument>())
+      return;
+    if (auto node = dyn_cast<NodeOp>(a.getDefiningOp()))
+      return getProductTerms(node.input(), terms);
+
+    if (auto src = getConnectSrc(a))
+      return getProductTerms(src, terms);
+
+    if (auto op = a.getDefiningOp()) {
+      if (auto andOp = dyn_cast<AndPrimOp>(op)) {
+        for (auto i : op->getOperands())
+          getProductTerms(i, terms);
+        return;
+      } else if (auto muxOp = dyn_cast<MuxPrimOp>(op)) {
+        if (ConstantOp cLow =
+                dyn_cast_or_null<ConstantOp>(muxOp.low().getDefiningOp()))
+          if (cLow.value().isZero()) {
+            getProductTerms(muxOp.sel(), terms);
+            getProductTerms(muxOp.high(), terms);
+            return;
+          }
+      }
+    }
+  }
+
+  bool checkComplement(SmallVector<Value> prodTerms[2]) {
+    bool isComplement = false;
+    for (auto t1 : prodTerms[0])
+      for (auto t2 : prodTerms[1]) {
+        if (!t1.isa<BlockArgument>() && isa<NotPrimOp>(t1.getDefiningOp()))
+          if (cast<NotPrimOp>(t1.getDefiningOp()).input() == t2)
+            return true;
```

Comment:
Ditto for the check below.
```suggestion
          return cast<NotPrimOp>(t1.getDefiningOp()).input() == t2;
```

Reply:
Actually, we don't want to return if this condition is false. We want the following code to execute, so the semantics is a bit different.